In [17]:
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import math

In [2]:
df=pd.read_csv('nhanes_2015_2016.csv')

In [5]:
df.head()

,SEQN,ALQ101,ALQ110,ALQ130,SMQ020,RIAGENDR,RIDAGEYR,RIDRETH1,DMDCITZN,DMDEDUC2,...,BPXDI2,BMXWT,BMXHT,BMXBMI,BMXLEG,BMXARML,BMXARMC,BMXWAIST,HIQ210,DMDMARTLx
0,83732,1.0,NaN,1.0,1,1,62,3,1.0,5.0,...,64.0,94.8,184.5,27.8,43.3,43.6,35.9,101.1,2.0,Married
1,83733,1.0,NaN,6.0,1,1,53,3,2.0,3.0,...,88.0,90.4,171.4,30.8,38.0,40.0,33.2,107.9,NaN,Divorced
2,83734,1.0,NaN,NaN,1,1,78,3,1.0,3.0,...,44.0,83.4,170.1,28.8,35.6,37.0,31.0,116.5,2.0,Married
3,83735,2.0,1.0,1.0,2,2,56,3,1.0,5.0,...,68.0,109.8,160.9,42.4,38.5,37.7,38.3,110.1,2.0,Cohabitating
4,83736,2.0,1.0,1.0,2,2,42,4,1.0,4.0,...,54.0,55.2,164.9,20.3,37.4,36.0,27.2,80.4,2.0,Divorced


## Question 1

Relabel the marital status variable [DMDMARTL](https://wwwn.cdc.gov/Nchs/Nhanes/2015-2016/DEMO_I.htm#DMDMARTL) to have brief but informative character labels.  Then construct a frequency table of these values for all people, then for women only and for men only.  Then construct these three frequency tables using only people whose age is between 30 and 40.

In [3]:
r = {1: "Married", 2: "Widowed", 3: "Divorced", 4: "Separated", 5: "NeverMarried", 6: "Cohabitating", 77: "Refused", 99: "Unknown"}
df["DMDMARTLx"] = df["DMDMARTL"].replace(r)

In [ ]:
print("All Subject")
x=df['DMDMARTLx'].value_counts()  # maritial status
print(x/x.sum())

All Subject
DMDMARTLx
Married         0.507855
NeverMarried    0.183412
Divorced        0.105773
Cohabitating    0.096273
Widowed         0.072342
Separated       0.033979
Refused         0.000365
Name: count, dtype: float64


In [7]:
df['RIAGENDRx']=df['RIAGENDR'].replace({1:"Male",2:"Female"})
print('All Gender')
y=df['RIAGENDRx'].value_counts()
print(y/y.sum())

All Gender
RIAGENDRx
Female    0.518919
Male      0.481081
Name: count, dtype: float64


In [13]:
for ky,db in df.groupby("RIAGENDRx"):
    print("\nRIAGENDR=",ky)
    x=db['DMDMARTLx'].value_counts()
    print(x/x.sum())


RIAGENDR= Female
DMDMARTLx
Married         0.457193
NeverMarried    0.182456
Divorced        0.122807
Widowed         0.103860
Cohabitating    0.091930
Separated       0.041404
Refused         0.000351
Name: count, dtype: float64

RIAGENDR= Male
DMDMARTLx
Married         0.562881
NeverMarried    0.184451
Cohabitating    0.100991
Divorced        0.087271
Widowed         0.038110
Separated       0.025915
Refused         0.000381
Name: count, dtype: float64


In [16]:
da3040=df.query('RIDAGEYR>=30 & RIDAGEYR<=40')
for ky,db in da3040.groupby('RIAGENDRx'):
    print('\nRIAGENDR=',ky,'30<=RIDAGEYR<=40')
    x=db['DMDMARTLx'].value_counts()
    print(x/x.sum())


RIAGENDR= Female 30<=RIDAGEYR<=40
DMDMARTLx
Married         0.535714
NeverMarried    0.218045
Cohabitating    0.122180
Divorced        0.086466
Separated       0.033835
Widowed         0.003759
Name: count, dtype: float64

RIAGENDR= Male 30<=RIDAGEYR<=40
DMDMARTLx
Married         0.556680
NeverMarried    0.204453
Cohabitating    0.157895
Divorced        0.048583
Separated       0.024291
Widowed         0.006073
Refused         0.002024
Name: count, dtype: float64


__Q1a.__ Briefly comment on some of the differences that you observe between the distribution of marital status between women and men, for people of all ages.

Compared to men, women have a lower probability of being currently married, a greater probability of being divorced, and a much greater probability of being widowed compared to men.

__Q1b.__ Briefly comment on the differences that you observe between the distribution of marital status states for women between the overall population, and for women between the ages of 30 and 40.

Women between the ages of 30 and 40 have a much lower probability of being widowed, a lower probability of being divorced, and a greater probability of cohabitating, compared to women of all ages.

__Q1c.__ Repeat part b for the men.

Men between the ages of 30 and 40 have a greater probability of cohabitating, a lower probability of being divorced, and a much lower probability of being widowed, compared to men of all ages.

## Question 2

Restricting to the female population, stratify the subjects into age bands no wider than ten years, and construct the distribution of marital status within each age band.  Within each age band, present the distribution in terms of proportions that must sum to 1.

In [21]:
# restrict the dataset to female only
da_mod=df.query("RIAGENDR==2").copy()
# calculate the minimum and maximum ages
minage=min(da_mod['RIDAGEYR'])
maxage=max(da_mod['RIDAGEYR'])


In [22]:
# one way to create the list of bins so that the first bin is closed on the left
bins = [minage] + list(range(int(math.ceil(minage/10)) * 10, int(math.ceil(maxage/10)) * 10, 10)) + [maxage]

# another way to create the list of bins without using the math library
# bins = [18, 20, 30, 40, 50, 60, 70, 80]
#print(bins)

# create labels for the bins
labels = [f"({bins[i]}, {bins[i+1]}]" if(i) else f"[{bins[i]}, {bins[i+1]}]" for i in range(len(bins)-1)]

# create the agegrp "bins" variable
da_mod["agegrp"] = pd.cut(da_mod.RIDAGEYR, bins, labels=labels, include_lowest = True)

# what does agegroup look like?
#print(da_mod.agegrp.value_counts(dropna=False))

# what does DMDMARTLx look like? Any rare or missing values?
#print(da_mod.DMDMARTLx.value_counts(dropna=False))

# distribution of marital status
dx = da_mod.loc[~da_mod.DMDMARTLx.isin(["Refused", np.nan]), :]  # drop rare/missing values
dx = dx.groupby(["agegrp"])["DMDMARTLx"]
dx = dx.value_counts()
dx = dx.unstack() # Restructure the results from 'long' to 'wide'
dx = dx.apply(lambda x: x/x.sum(), axis=1) # Normalize within each stratum to get proportions
print("Females")
print(dx.to_string(float_format="%.3f"))  # Limit display to 3 decimal places

# alternative that replaces NaN with 0
#print(dx.fillna(0).to_string(float_format="%.3f"))

Females
DMDMARTLx  Cohabitating  Divorced  Married  NeverMarried  Separated  Widowed
agegrp                                                                      
[18, 20]          0.205       NaN    0.026         0.769        NaN      NaN
(20, 30]          0.206     0.021    0.305         0.446      0.021      NaN
(30, 40]          0.120     0.091    0.544         0.205      0.036    0.004
(40, 50]          0.074     0.137    0.574         0.125      0.066    0.024
(50, 60]          0.068     0.177    0.548         0.090      0.058    0.060
(60, 70]          0.043     0.193    0.481         0.086      0.050    0.147
(70, 80]          0.007     0.144    0.317         0.051      0.020    0.461


__Q2a.__ Comment on the trends that you see in this series of marginal distributions.

##### 1. Cohabitating
* it is most common among younger adults and declines with age
* Around 2-4% of female under 30 are cohabiting

##### 2. Divorced
* divorced is rare among the youngest females(at age 18-30)
* it increases age age increases for 30-70 
* it decrease approx 5% after 70

##### 3. Married
* married rate increases as age for 18-60
* married rate decreases after age 60-80

##### 4. NeverMarried
* Never married decreases sharply with age
* it is at peak for age group 18-20 

##### 5. Separated
* separation increases for age group 20-50
* most female separated at age group 40-50 i.e 6%

##### 6. Widowed
* it is absent for age group 18-30
* it increases with age group
* it is peak at age group 70-80



__Q2b.__ Repeat the analysis for males.

In [23]:
# restrict the dataset to male only
da_mod=df.query("RIAGENDR==1").copy()
# calculate the minimum and maximum ages
minage=min(da_mod['RIDAGEYR'])
maxage=max(da_mod['RIDAGEYR'])


In [25]:
# one way to create the list of bins so that the first bin is closed on the left
bins = [minage] + list(range(int(math.ceil(minage/10)) * 10, int(math.ceil(maxage/10)) * 10, 10)) + [maxage]

# another way to create the list of bins without using the math library
# bins = [18, 20, 30, 40, 50, 60, 70, 80]
#print(bins)

# create labels for the bins
labels = [f"({bins[i]}, {bins[i+1]}]" if(i) else f"[{bins[i]}, {bins[i+1]}]" for i in range(len(bins)-1)]

# create the agegrp "bins" variable
da_mod["agegrp"] = pd.cut(da_mod.RIDAGEYR, bins, labels=labels, include_lowest = True)

# what does agegroup look like?
#print(da_mod.agegrp.value_counts(dropna=False))

# what does DMDMARTLx look like? Any rare or missing values?
#print(da_mod.DMDMARTLx.value_counts(dropna=False))

# distribution of marital status
dx = da_mod.loc[~da_mod.DMDMARTLx.isin(["Refused", np.nan]), :]  # drop rare/missing values
dx = dx.groupby(["agegrp"])["DMDMARTLx"]
dx = dx.value_counts()
dx = dx.unstack() # Restructure the results from 'long' to 'wide'
dx = dx.apply(lambda x: x/x.sum(), axis=1) # Normalize within each stratum to get proportions
print("males")
print(dx.to_string(float_format="%.3f"))  # Limit display to 3 decimal places

# alternative that replaces NaN with 0
#print(dx.fillna(0).to_string(float_format="%.3f"))

males
DMDMARTLx  Cohabitating  Divorced  Married  NeverMarried  Separated  Widowed
agegrp                                                                      
[18, 20]          0.075       NaN    0.025         0.900        NaN      NaN
(20, 30]          0.213     0.005    0.238         0.523      0.016    0.005
(30, 40]          0.158     0.053    0.565         0.195      0.026    0.004
(40, 50]          0.082     0.085    0.703         0.097      0.027    0.005
(50, 60]          0.075     0.126    0.652         0.104      0.022    0.022
(60, 70]          0.050     0.126    0.666         0.087      0.032    0.039
(70, 80]          0.022     0.142    0.612         0.022      0.035    0.167


##### 1. Cohabitating
* it is peak for age group 20-30 i.e 21%
* it is decreasing significantly as age decline

##### 2. Divorced
* divorce is peak at age group 70-80 
* divorce increases age increases 
* it is lowest at age 18-20

##### 3. Married
* married rate increases as age increase
* married rate is highest at age 40-50
* married rate decreases after age 60-80

##### 4. NeverMarried
* Never married decreases sharply with age
* it is at peak for age group 18-20 

##### 5. Separated
* separation increases for age group 20-50
* most male separated at age group 70-80

##### 6. Widowed
* it is absent for age group 18-30
* it increases with age group
* it is peak at age group 70-80



__Q2c.__ Comment on any notable differences that you see when comparing these results between females and for males.